xarray/core/weighted.py

```diff
@@ -0,0 +1,252 @@
+from .computation import dot
+from typing import TYPE_CHECKING, Hashable, Iterable, Optional, Tuple, Union, overload
+
+if TYPE_CHECKING:
+    from .dataarray import DataArray, Dataset
+
+_WEIGHTED_REDUCE_DOCSTRING_TEMPLATE = """
+    Reduce this {cls}'s data by a weighted `{fcn}` along some dimension(s).
+
+    Parameters
+    ----------
+    dim : str or sequence of str, optional
+        Dimension(s) over which to apply the weighted `{fcn}`.
+    skipna : bool, optional
+        If True, skip missing values (as marked by NaN). By default, only
+        skips missing values for float dtypes; other dtypes either do not
+        have a sentinel missing value (int) or skipna=True has not been
+        implemented (object, datetime64 or timedelta64).
+        Note: Missing values in the weights are replaced with 0 (i.e. no
+        weight).
+    keep_attrs : bool, optional
+        If True, the attributes (`attrs`) will be copied from the original
+        object to the new one.  If False (default), the new object will be
+        returned without attributes.
+    **kwargs : dict
+        Additional keyword arguments passed on to the appropriate array
+        function for calculating `{fcn}` on this object's data.
+
+    Returns
+    -------
+    reduced : {cls}
+        New {cls} object with weighted `{fcn}` applied to its data and
+        the indicated dimension(s) removed.
+    """
+
+_SUM_OF_WEIGHTS_DOCSTRING = """
+    Calculate the sum of weights, accounting for missing values
+
+    Parameters
+    ----------
+    dim : str or sequence of str, optional
+        Dimension(s) over which to sum the weights.
+
+    Returns
+    -------
+    reduced : {cls}
+        New {cls} object with the sum of the weights over the given dimension.
+
+
+    """
+
+
+def _maybe_get_all_dims(
+    dims: Optional[Union[Hashable, Iterable[Hashable]]],
+    dims1: Tuple[Hashable, ...],
+    dims2: Tuple[Hashable, ...],
+):
+    """ the union of dims1 and dims2 if dims is None
+
+    `dims=None` behaves differently in `dot` and `sum`, so we have to apply
+    `dot` over the union of the dimensions
```

Comment:
I'm sure you're right, but can you help me understand the issue here? Doesn't each of `dot` and `sum` collapse all dimensions if `dims=None`? Or there's an issue where only a subset of the dimensions are shared?